fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,1170 @@
+    <https://dl.acm.org/doi/pdf/10.1145/3278721.3278779>`_ [1]_.
+    This algorithm takes as input two neural network
+    models, a predictor model and an adversarial model, defined either as a
+    `PyTorch module
+    <https://pytorch.org/docs/stable/generated/torch.nn.Module.html>`_ or
+    `Tensorflow2 model
+    <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`_. You train this
+    predictor using an API that is similar to estimators in `sklearn`.
+
+    As per [1]_, the neural network models :code:`predictor_model` and
+    :code:`adversary_model` should not have a discrete prediction at the end of the
+    model, as gradients do not propagate through a discretization step.
+    When discrete predictions are required, as is the case in binary and
+    multiclass classification, please just supply the predictor model that
+    produces continuous values, e.g., using a sigmoidal or soft-max
+    activation, and specify the appropriate final transformation separately via
+    parameter :code:`prediction_function` (or let us automatically infer this).
+
+    We only support single-feature predictions. Sensitive features must be
+    single-feature discrete, or (possibly multi-output) continuous.
+    We apply preprocessing to ensure everything is encoded as floats, and
+    appropriate loss functions/predictor functions are inferred.
+
+    Parameters
+    ----------
+    backend : str, BackendEngine, default = 'auto'
+        The backend to use. Must be one of :code:`'torch'`, :code:`'tensorflow'`,
+        or :code:`'auto'` which indicates PyTorch, TensorFlow, or to
+        automatically infer
+        the backend from the :code:`predictor_model` and which are installed.
+        You can also pass in a BackendEngine class.
+
+    predictor_model : list, torch.nn.Module, tensorflow.keras.Model
+        The predictor model to train.
+        Instead of a neural network model, we can pass a list of keywords
+        :math:`k_1, k_2, \dots` that indicate either
+        the number of nodes :math:`k_i` (if :math:`k_i` is integer) or a keyword
+        that indicates an activation function (if :math:`k_i` is a string) or
+        a layer or activation function instance directly (if :math:`k_i` is
+        callable). The default parameter is :code:`[]`, which indicates
+        a neural network without any hidden layers.
+        However, the number of nodes in the input
+        and output layer are automatically inferred from data, and the final
+        activation function (such as softmax for categorical
+        predictors) are inferred from data.
+        If :code:`backend` is specified, you cannot pass a model
+        that uses a different backend.
+
+    adversary_model : list, torch.nn.Module, tensorflow.keras.Model
+        The adversary model to train. Defined similarly as predictor_model.
+        Must be the same type as the
+        :code:`predictor_model`.
+
+    predictor_function : str, callable, default = 'auto'
```

Comment:
I think we should remove this parameter (see [this comment](https://github.com/fairlearn/fairlearn/pull/973#discussion_r851659798)). Since we always default to either binary or multinomial logistic loss, the predictor function is always determined from those.